spec/classes/chrony_spec.rb

```diff
@@ -26,25 +42,32 @@
         case facts[:osfamily]
         when 'Archlinux'
           context 'using defaults' do
-            it { is_expected.to contain_file('/etc/chrony.conf').with_content(%r{^\s*cmdallow 127\.0\.0\.1$}) }
-            ['0.pool.ntp.org', '1.pool.ntp.org', '2.pool.ntp.org', '3.pool.ntp.org'].each do |s|
-              it { is_expected.to contain_file('/etc/chrony.conf').with_content(%r{^\s*server #{s} iburst$}) }
+            it do
+              is_expected.to contain_file(config_file)
+                .with_content(%r{^\s*cmdallow 127\.0\.0\.1$})
+                .with_content(%r{^\s*server 0.pool.ntp.org iburst$})
+                .with_content(%r{^\s*server 1.pool.ntp.org iburst$})
+                .with_content(%r{^\s*server 2.pool.ntp.org iburst$})
+                .with_content(%r{^\s*server 3.pool.ntp.org iburst$})
+                .with_content(%r{^\s*rtconutc$})
+                .with_content(%r{^\s*driftfile /var/lib/chrony/drift$})
+                .with_content(%r{^\s*rtcsync$})
+                .with_content(%r{^\s*dumpdir /var/lib/chrony$})
+                .without_content(%r{^\s*\n\s*$})
+            end
+            it do
+              is_expected.to contain_file(keys_file)
```

Comment:
Given these are now the same, you can pull it outside of the `case facts[:osfamily]` statement to reduce the duplication further. I'd suggest to also move `context 'using defaults'` outside of the case statement. Then the case statement is just for the `config_file`.

Reply:
If you mean replace the blocks

```ruby
          it do
            is_expected.to contain_file(keys_file)
              .with_mode('0644')
              .with_owner('0')
              .with_group('0')
              .with_replace(true)
              .with_content("0 xyzzy\n")
          end
```

with a single block, that's not possible since the `mode` differs between osfamilies.

Reply:
So this is sort of where it may break down. If you copy the block, a variable for `keys_file` isn't that useful. You can either add a `keys_file_mode` variable. Another is to question whether they really need to differ. For example, I wonder if a keys file should be world readable.

Reply:
Yes the mode should probably default to 0640 for all osfamilies. But I think it kind of still make sense to have a case statement for both the keyfile and configfile for the ```using defaults``` test, since the defaults is in fact set by param.pp different on each osfamily, and the might change some time.  The ```with some params passed in``` on the other hand don't need to.

Reply:
One possible solution:

```ruby
context 'using defaults' do
  it do
    is_expected.to contain_file(keys_file)
              .with_owner('0')
              .with_group('0')
              .with_replace(true)
              .with_content("0 xyzzy\n")
  end

  case facts[:os_family]
  when 'ArchLinux'
    # config_file here
    it { is_expected.to contain_file(keys_file).with_mode('0644') }
  # other distros
  end
end
```

That way you can clearly see the common parts and those that differ. Now this may be slower if it needs to be compiled multiple times (@alexjfisher suggested it doesn't, haven't benchmarked myself, only feels like it).

Of course unifying them all as much as possible is the easiest for maintenance.

Reply:
Do you mean like ad65d226842eb72ed8de6b8ad674a6500b097df7? No that didn't work either, also group differs between osfamilies.